lib: use explicit name imports in `lib/generators.nix`

philiptaron opened the pull request:
## Description of changes

Using `with` at a top-level scope is [an anti-pattern](https://nix.dev/guides/best-practices#with-scopes). The tracking issue is https://github.com/NixOS/nixpkgs/issues/208242. This is a pure refactor: there is, or ought to be, no functional change contained in this PR.

This PR is best reviewed commit-by-commit.

Related: 
- https://github.com/NixOS/nixpkgs/pull/293901
- https://github.com/NixOS/nixpkgs/pull/295374

## Things done

- [x] Tested, as applicable:
  - [x] `nix-instantiate --strict --eval lib/tests/misc.nix`
  - [x] `lib/tests/modules.sh`
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

philiptaron committed:
lib/generators: builtins.isFloat is in Nix 2.3

There's a couple of aliases in play:

1. `lib.isList` is an alias for `builtins.isList`.
2. `lib.strings.concatStringsSep` is an alias for `builtins.concatStringsSep`

philiptaron committed:
lib/generators: explicitly import names from `lib`

I followed the `inherit` chains in `lib/default.nix` to arrive at these imports.

philiptaron committed:
lib/generators: explicitly import names from `lib.string`

A couple of these were imports from `builtins`.

philiptaron committed:
lib/generators: explicitly import names from `lib.attrsets`

Everything used was already imported.

philiptaron committed:
lib/generators: explicitly import names from `lib.trivial`

Everything else was already imported.

philiptaron committed:
lib/generators: use the explicit public interface pattern

This enables further refactoring without accidentally changing the public interface.

philiptaron commented on lib/generators.nix:
```diff
@@ -14,29 +14,84 @@
-rec {
+let
+  inherit (builtins)
+    addErrorContext
+    attrNames
+    concatLists
+    concatStringsSep
+    elem
+    filter
+    head
+    isAttrs
+    isBool
+    isFloat
+    isInt
+    isList
+    isPath
+    isString
+    length
+    mapAttrs
+    match
+    replaceStrings
+    split
+    tail
+    toJSON
+    typeOf
+    ;
```
Of these names, all but `match`, `split`, `toJSON`, and `typeof` could be imported from `lib` directly. Is there a preference?

infinisil replied:
I'd prefer `lib` be used whenever possible. `lib` should ideally expose all `builtins`.

philiptaron replied:
Done.

AndersonTorres replied:
> lib should ideally expose all builtins.

Nice. I will abuse this from now on :)

infinisil commented on lib/generators.nix:
```diff
@@ -542,4 +578,37 @@ ${expr "" v}
+
+in
+
+# Everything in this attrset is the public interface of the file.
+{
+  inherit
+    mkDconfKeyValue
+    mkKeyValueDefault
```
This could be a problem if we try to render documentation from this file in the future (like with other sublibraries). I don't think nixdoc would look at documentation above `inherit` identifiers.

Not a problem for now though. Overall the pattern of being more explicit about what is being exported sounds good to me.

philiptaron replied:
Turns out that `nixdoc` sees through these lines without an issue. Opened https://github.com/nix-community/nixdoc/pull/111 to rework a test to show that's the case.

fricklerhandwerk reviewed (approved):
Looks good, modulo https://github.com/NixOS/nixpkgs/pull/293835#pullrequestreview-1929303310

infinisil commented on lib/generators.nix:
```diff
@@ -14,29 +14,81 @@
 
-rec {
+let
+  inherit (lib)
+    addErrorContext
+    assertMsg
+    attrNames
```
What's the reason behind using `lib` only for some fields but `lib.<sublib>` for others? E.g. `attrNames` comes from `lib`, but `isDerivation` comes from `lib.attrsets`, even though they're available in both.

philiptaron replied:
This was the first PR I made against `lib`, so the _historical_ reason is that I was trying to do the work by hand and I didn't have the script I later wrote to gather and use the names automatically.

If you have a preference, I'd be glad to accommodate.

infinisil replied:
Discussed [on Matrix](https://matrix.to/#/!djTaTBQyWEPRQxrPTb:nixos.org/$jWTFrysx9Y8swFg5ejBb_ba2eVpoFgwVAgb5YesCcqI?via=nixos.org&via=matrix.org&via=nixos.dev), let's use the ones from `lib`. While I did do it the other way around in [`lib.fileset`](https://github.com/NixOS/nixpkgs/blob/master/lib/fileset/default.nix#L101), I think it should just be consistent within the same file for now.

philiptaron replied:
Done.

philiptaron committed:
lib: use names from `lib` in `lib/generators.nix`, rather than `builtins` or submodules of `lib`

There's not a lot of rhyme to which names are exported from which module,
as I see it, but everything is found somewhere.

infinisil commented on lib/generators.nix:
```diff
@@ -14,29 +14,73 @@
+    flatten
+    foldl
+    functionArgs  # Note: not the builtin; considers `__functor` in attrsets.
+    generators
+    gvariant
```
`generators` is not used :)

```suggestion
    gvariant
```

`gvariant` here is a bit out of place since it's a sublibrary with nested attributes, but I don't see a problem with it.

philiptaron replied:
> gvariant here is a bit out of place since it's a sublibrary with nested attributes, but I don't see a problem with it.

It's just a name 🤷🏻 As previously mentioned, the rhyme is a little hard to find.
